cleverhans/evaluation.py

```diff
@@ -0,0 +1,338 @@
+# `graph_factory`, to avoid making the tf graph too big
+_batch_eval_multi_worker_cache = {}
+
+
+def batch_eval_multi_worker(sess, graph_factory, numpy_inputs, batch_size=None,
+                            devices=None, feed=None):
+    """
+    Generic computation engine for evaluating an expression across a whole
+    dataset, divided into batches.
+
+    This function assumes that the work can be parallelized with one worker
+    device handling one batch of data. If you need multiple devices per
+    batch, use `batch_eval`.
+
+    The tensorflow graph for multiple workers is large, so the first few
+    runs of the graph will be very slow. If you expect to run the graph
+    few times (few calls to `batch_eval_multi_worker` that each run few
+    batches) the startup cost might dominate the runtime, and it might be
+    preferable to use the single worker `batch_eval` just because its
+    startup cost will be lower.
+
+    :param sess: tensorflow Session
+    :param graph_factory: callable
+        When called, returns (tf_inputs, tf_outputs) where:
+            tf_inputs is a list of placeholders to feed from the dataset
+            tf_outputs is a list of tf tensors to calculate
+        Example: tf_inputs is [x, y] placeholders, tf_outputs is [accuracy].
+        This factory must make new tensors when called, rather than, e.g.
+        handing out a reference to existing tensors.
+        This factory must make exactly equivalent expressions every time
+        it is called, otherwise the results of `batch_eval` will vary
+        depending on how work is distributed to devices.
+        This factory must respect "with tf.device()" context managers
+        that are active when it is called, otherwise work will not be
+        distributed to devices correctly.
+    :param numpy_inputs:
+        A list of numpy arrays defining the dataset to be evaluated.
+        The list should have the same length as tf_inputs.
+        Each array should have the same number of examples (shape[0]).
+        Example: numpy_inputs is [MNIST().x_test, MNIST().y_test]
+    :batch_size: Number of examples to use in a single evaluation batch.
+        If not specified, this function will use a reasonable guess and
+        may run out of memory.
+        When choosing the batch size, keep in mind that the batch will
+        be divided up evenly among available devices. If you can fit 128
+        examples in memory on one GPU and you have 8 GPUs, you probably
+        want to use a batch size of 1024 (unless a different batch size
+        runs faster with the ops you are using, etc.)
+    :param devices: List of devices to run on. If unspecified, uses all
+        available GPUs if any GPUS are available, otherwise uses CPUs.
+    :param feed: An optional dictionary that is appended to the feeding
+             dictionary before the session runs. Can be used to feed
+             the learning phase of a Keras model for instance.
+    """
```

Comment:
:returns: List of numpy arrays corresponding the outputs prodced by the graph_factory?

Reply:
Thanks, I added this to the docstring.